// prismarine-nbt/sample/bigtest.js
module.exports = {
  name: 'Level',
  type: 'compound',
  value: {
    longTest: {
      type: 'long',
      value: [
        2147483647,
        -1
      ]
    },
    shortTest: {
      type: 'short',
      value: 32767
    },
    stringTest: {
      type: 'string',
      value: 'HELLO WORLD THIS IS A TEST STRING ÅÄÖ!'
    },
    floatTest: {
      type: 'float',
      value: 0.4982314705848694
    },
    intTest: {
      type: 'int',
      value: 2147483647
    },
    'nested compound test': {
      type: 'compound',
      value: {
        ham: {
          type: 'compound',
          value: {
            name: {
              type: 'string',
              value: 'Hampus'
            },
            value: {
              type: 'float',
              value: 0.75
            }
          }
        },
        egg: {
          type: 'compound',
          value: {
            name: {
              type: 'string',
              value: 'Eggbert'
            },
            value: {
              type: 'float',
              value: 0.5
            }
          }
        }
      }
    },
    'listTest (long)': {
      type: 'list',
      value: {
        type: 'long',
        value: [
          [
            0,
            11
          ],
          [
            0,
            12
          ],
          [
            0,
            13
          ],
          [
            0,
            14
          ],
          [
            0,
            15
          ]
        ]
      }
    },
    'listTest (compound)': {
      type: 'list',
      value: {
        type: 'compound',
        value: [
          {
            name: {
              type: 'string',
              value: 'Compound tag #0'
            },
            'created-on': {
              type: 'long',
              value: [
                294,
                1379390861
              ]
            }
          },
          {
            name: {
              type: 'string',
              value: 'Compound tag #1'
            },
            'created-on': {
              type: 'long',
              value: [
                294,
                1379390861
              ]
            }
          }
        ]
      }
    },
    byteTest: {
      type: 'byte',
      value: 127
    },
    'byteArrayTest (the first 1000 values of (n*n*255+n*7)%100, starting with n=0 (0, 62, 34, 16, 8, ...))': {
      type: 'byteArray',
      value: [
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48,
        0,
        62,
        34,
        16,
        8,
        10,
        22,
        44,
        76,
        18,
        70,
        32,
        4,
        86,
        78,
        80,
        92,
        14,
        46,
        88,
        40,
        2,
        74,
        56,
        48,
        50,
        62,
        84,
        16,
        58,
        10,
        72,
        44,
        26,
        18,
        20,
        32,
        54,
        86,
        28,
        80,
        42,
        14,
        96,
        88,
        90,
        2,
        24,
        56,
        98,
        50,
        12,
        84,
        66,
        58,
        60,
        72,
        94,
        26,
        68,
        20,
        82,
        54,
        36,
        28,
        30,
        42,
        64,
        96,
        38,
        90,
        52,
        24,
        6,
        98,
        0,
        12,
        34,
        66,
        8,
        60,
        22,
        94,
        76,
        68,
        70,
        82,
        4,
        36,
        78,
        30,
        92,
        64,
        46,
        38,
        40,
        52,
        74,
        6,
        48
      ]
    },
    doubleTest: {
      type: 'double',
      value: 0.4931287132182315
    }
  }
}
